### ISSUE TITLE
Mapping a Long to a GraphQL Int (instead of a BigInteger scalar)

### ISSUE THREAD

**Issue by marceloverdijk** [eclipse/microprofile-graphql#260]
Java `Long`'s are automatically converted to the `BigInteger` scalar - which is added by M GraphQL.

I would prefer to map it to the _standard_ GraphQL `Int` scalar instead, is that possible to configure (and if so configure it globally)?

**Comment by phillip-kruger** [comment]
It's out of range to be mapped to an Int. No ?

**Comment by marceloverdijk** [comment]
For a Java `Integer` yes, but not for a GraphQL `Int`.

GraphQL has only `Int` and `Float` (and no `Long` or `Double` as the GraphQL spec does not specify a maximum byte width or storage size for its `Int` and `Float` types)

So internally (GraphQL @Type classes in Java) you must use a `Long` in case it a big int, but in the exposed GraphQL api it's more than fine to use the GraphQL `Int` type.

Thats what I always did so far using plain GraphQL-java lib.
Prototyping with Quarkus-GraphQL I noticed my `Long` were exposed as `BigInteger`.

**Comment by phillip-kruger** [comment]
From the Spec: https://spec.graphql.org/June2018/#sec-Int:

"The Int scalar type represents a signed 32‐bit numeric non‐fractional value. Response formats that support a 32‐bit integer or a number type should use that type to represent this scalar."

For Java: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html
Java Long is a 64 bit.

Also from the spec:
"A server may omit any of the built‐in scalars from its schema, for example if a schema does not refer to a floating‐point number, then it must not include the Float type. However, if a schema includes a type with the name of one of the types described here, it must adhere to the behavior described. As an example, a server must not include a type called Int and use it to represent 128‐bit numbers, internationalization information, or anything other than what is defined in this document."

This is how we interpreted it when we decided to map Long to BigInteger.

I would say if you want to have Int in GraphQL then use Integer in Java?

**Comment by marceloverdijk** [comment]
Good catch in the spec.

I looked at this one in the past:

https://github.com/graphql/graphql-spec/issues/73#issuecomment-129268000

@leebyron:

> The GraphQL spec does not specify a maximum byte width or storage size for its Int and Float type, but there is a note about the caveat of platform specific restrictions. The two types only represent the semantic difference between whole and part numbers.

> This means that you can use Int as a Long or Float as a Double as long as the platforms you're using support that size.

> GraphQL allows you to define custom scalars as well, in the case that you want to distinguish between Long and Int, or define a custom serialization for it.

As it is an old comment, maybe the spec was updated later.


Looking at the Github GraphQL docs I see they e,g, use a GraphQL `Int` for the `totalCount` field (which is typically a `Long` in the Java world:

![image](https://user-images.githubusercontent.com/197004/84476551-9bf3c880-ac8e-11ea-97fa-c8c4d9e9dcf2.png)

**Comment by phillip-kruger** [comment]
Still, even if GitHub have done it, we can not assume it. What we can assume that if you use a Long and not an Integer in Java, you expect the use to be greater than the range specified by Integer, so we can not map to an Integer. We can _maybe_ allow configuration to allow that, but I am not so sure if we should do that. (If you can configure it, you might as well use an Integer in Java ?) 

What does the others think ?
@andymc12 @kaqqao @ybroeker @WhiteCat22 @tmiddlet2666 @jefrajames @thejibz

**Comment by marceloverdijk** [comment]
Indeed it should not be assumed, a configuration option is exactly what I meant :-)

**Comment by andymc12** [comment]
I agree that default behavior should map Java `int`/`Integer` types to GraphQL `Int`, and Java `long`/`Long` to GraphQL (custom scalar) `BigInteger` for maximum compatibility.  But I don't object to a switch (either in the spec or the implementation) that would map Java `long`/`Long` types to GraphQL `Int`.  

This sounds like it could be another flavor of the discussion we've had about mapping scalars - in the case @t1 presented, it was more of mapping custom Java types to GraphQL `String` types, but maybe we could re-use the mapping solution to allow Java `long`/`Long` to GraphQL `Int` as well?

**Comment by phillip-kruger** [comment]
I like that @andymc12 ! So should we bake it in the implementation with the already custom mapping here: https://github.com/smallrye/smallrye-graphql/issues/249 then we move it up to spec if we are happy with it ? And what we are saying is that that mapping will not only allow you to map a non-mapped Java type to an existing Scalar, but also map an already mapped Java type to another Scalar ? Something like that ?

**Comment by phillip-kruger** [comment]
This is now supported (per field) in SmallRye by using the `@ToScalar` annotation. We will keep this issue open until we move that up to the spec.

**Comment by Andro999b** [comment]
@phillip-kruger i think it also relevant to Double. Double should be mapped as BigDecimal. Currently when i use double value rounded to 5 numeral after decimal point(that not relevant for float). And yes `ToScalar` fix it.

**Comment by phillip-kruger** [comment]
Yes double maps to float (see https://github.com/smallrye/smallrye-graphql/blob/b69fe8649f880147b5bb7619d0157f1aa204e0bc/server/implementation/src/main/java/io/smallrye/graphql/scalar/number/FloatScalar.java#L32)

Are you suggesting it should rather map to BigDecimal ?

**Comment by Andro999b** [comment]
@phillip-kruger Yes in current implementation double response loos precision in response(that not happen for float). I think its because `.floatValue()` conversion(49.9902794 with floatValue() will round it to 49.99028). But all fine with BigDecimal.

**Comment by phillip-kruger** [comment]
Ok make sense, maybe we can fix it in the implementation. For now you can use @ToScalar. Can you please open an issue in SmallRye GraphQL. Thanks

**Comment by Andro999b** [comment]
Done https://github.com/smallrye/smallrye-graphql/issues/727

**Comment by phillip-kruger** [comment]
Thanks :)
